### PR TITLE
[11.0][ADD] tourist tax

### PR DESCRIPTION
**Pull request by mclaeysb**
Adds Tourist Tax to Hotel Folio.

- Taxes can be applied automatically when creating and editing folios
- Taxed amounted can be computed at flat rate, per night and per person (counting adults and/or children).

Taxes are added as a Hotel Service line to folios with the amount in the price. This is certainly not the cleanest approach (since they are not in a place of their own and since they are not accounting-wise processed as a tax), but the best/easiest way we could think of. Making a new object similar to service lines seemed like to much work (given all functions in the `hotel` module where services show up), and designing a tax with a fixed (i.e. not %) but variable price is not possible.

Question: I couldn't wrap the `update_tourist_tax` inside on `@api.onchange` because it should act on the record, not a temporary copy. Therefore I wrapped it in the write function, but this gave rise to triggers in stage cases (see the note on recursion: a write triggers a new write). Any tips on how to do this better/cleaner?

**Please do not merge before confirmation**

### CODE REVIEW

File: hotel_reservation/models/hotel_reservation.py [hunk 1]
```diff
@@ -545,6 +545,8 @@ def create_folio(self):
                 "partner_shipping_id": reservation.partner_shipping_id.id,
                 "checkin_date": reservation.checkin,
                 "checkout_date": reservation.checkout,
+                "adults": reservation.adults,
+                "children": reservation.children,
                 "duration": duration,
```
Comment (robinkeunen) [inline hotel_reservation/models/hotel_reservation.py:550]: 👍

File: hotel_tourist_tax/models/hotel_folio.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            if (
+                bool(vals.get("duration"))
+                or bool(vals.get("adults"))
+                or bool(vals.get("children"))
+            ) and not bool(
+                vals.get("service_lines")  # prevent recursion
+                or rec.invoice_status == "invoiced"
+                or bool(rec.hotel_invoice_id)
+            ):
+                rec.update_tourist_tax()
```
Comment (robinkeunen) [inline hotel_tourist_tax/models/hotel_folio.py]: why not this formulation? (`in` keyword and useless boolean casting)

```py
if (
    "duration" in vals or "adults" in vals or "children" in vals
) and not (
    "service_lines" in vals 
    or rec.invoice_status == "invoiced" 
    or rec.hotel_invoice_id
):
```

Your version of the condition does not pass if `adults` and `children` are in `vals` and equals 0. But that might be what you want.

File: hotel_tourist_tax/models/hotel_folio.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        self.ensure_one()
+
+        if not self.order_id:
+            return
+
+        # Remove existing taxes
+        for service_line in self.service_lines:
+            if service_line.product_id in self.env["hotel.tourist.tax"].search(
+                []
+            ).mapped("hotel_service_id.product_id"):
```
Comment (robinkeunen) [inline hotel_tourist_tax/models/hotel_folio.py:44]: for clarity and performance (sql request are sent at each loop), you could prepare the tax product list outside the loop.

**Review by robinkeunen — APPROVED** [review]
Small question on a condition, otherwise lgtm

File: hotel_tourist_tax/models/hotel_folio.py (hunk 2 above)
Comment (mclaeysb) [reply hotel_tourist_tax/models/hotel_folio.py]: Aha, much cleaner. This statement catcher more cases indeed, but that's a good thing. Thanks!

**Pull request closed without merging**